core/server/common/src/main/java/alluxio/master/MasterRegistry.java

```diff
@@ -116,9 +95,56 @@ public MasterRegistry() {}
   /**
    * @return a collection of all the registered masters
    */
-  public synchronized Collection<Master> getMasters() {
-    List<Master> masters = new ArrayList<>(mRegistry.values());
-    Collections.sort(masters, COMPARATOR);
+  public Collection<Master> getMasters() {
+    List<Map.Entry<Class<?>, Master>> entries = new ArrayList<>(mRegistry.entrySet());
+    Collections.sort(entries, new DependencyComparator());
+    List<Master> masters = new ArrayList<>();
+    for (Map.Entry<Class<?>, Master> entry : entries) {
+      masters.add(entry.getValue());
+    }
     return masters;
   }
+
+  private Set<Class<?>> getTransitiveDeps(Class<?> clazz) {
```

Comment:
ah, the dependencies need to be converted to masters with the registry, changing
```java
  for (Master dep : queue.pop().getDependencies()) {
    ...
```
to
```java
  for (Class<?> depClass : queue.pop().getDependencies()) {
    Master dep = mRegistry.get(depClass);
    ...
```

Then we would be operating on Masters everywhere else